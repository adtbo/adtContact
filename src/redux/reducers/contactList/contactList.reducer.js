import { createAction } from 'redux-actions';
import typeToReducer from 'type-to-reducer';

import { getContactList, deleteContact } from '../../../services/contact/contact.service';

// Constants

const GET_CONTACTS = 'LIST/GET_CONTACTS';
const DELETE_CONTACT = 'LIST/DELETE_CONTACT';
const SET_SEARCH = 'LIST/SET_SEARCH';

export const actionTypes = {
    GET_CONTACTS,
    DELETE_CONTACT,
    SET_SEARCH,
};

// Actions

export const fetchContactList = createAction(GET_CONTACTS, getContactList);
export const deleteSelectedContact = createAction(DELETE_CONTACT, (contactId) => deleteContact(contactId));
export const setSearch = createAction(SET_SEARCH, (text) => Promise.resolve(text));

export const actions = {
    fetchContactList,
    deleteSelectedContact,
    setSearch,
};

// Reducer

// Initial State

export const initialState = {
    contacts: [],
    displayed: [],
    searchString: '',
    error: {
        status: false,
        message: '',
    },
}

const getContactsHandler = (state, action) => {
    
    const sortedContacts = action.payload.data.sort((a,b) => {
        let fa = (a.firstName+" "+a.lastName),
        fb = (b.firstName+" "+b.lastName);

        if(fa < fb) {
            return -1;
        }
        if(fa > fb){
            return 1;
        }
        else{
            return 0;
        }
    })

    return {
        ...state,
        contacts: sortedContacts,
        displayed: sortedContacts
    }
};

const errorHandler = (state, action) => {
    return {
        ...state,
        error:{
            status: true,
            message: action.payload.message
        }
    }
};

const setSearchHandler = (state, action) => {
    
    const filter = action.payload.toLowerCase();
    let newCon = [...state.contacts];
    let filteredContact = newCon.filter((data) => {
        return ((data.firstName+" "+data.lastName).toLowerCase()).indexOf(filter) >= 0;
    });
    
    return {
        ...state,
        displayed: filteredContact,
        searchString: action.payload,
    }
}

// type to reducers
export default typeToReducer({
    [GET_CONTACTS]: {
        FULFILLED: getContactsHandler,
        REJECTED: errorHandler
    },
    [SET_SEARCH]:{
        FULFILLED: setSearchHandler
    },
}, initialState);