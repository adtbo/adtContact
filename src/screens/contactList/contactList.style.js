import { Platform, StyleSheet } from 'react-native';

export default StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
    },
    contactItem:{
        flex:1,
        flexDirection: 'row',
        margin: 10,
        padding: 4,
        alignItems: 'center'
    },
    searchInput:{
        borderRadius: 40,
        paddingHorizontal: 10
    },
    searchCont:{
    },
    fabContainer: {
        justifyContent: 'flex-end',
        flexDirection: 'row',
        position: 'absolute',
        right: 20,
        bottom: 20
    },
    fabButton: {
        backgroundColor: '#E88C30',
        borderRadius: 35,
        width: 50,
        height: 50,
        alignItems: 'center',
        justifyContent: 'center'
    }
})