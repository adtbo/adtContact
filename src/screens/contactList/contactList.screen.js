import React, { useEffect, useLayoutEffect } from 'react'
import { Text, View, TouchableOpacity, FlatList, Image, Alert } from 'react-native'
import styles from './contactList.style'

import { connect } from 'react-redux'
import { actions } from '../../redux/reducers/contactList/contactList.reducer'
import { actions as detailActions } from '../../redux/reducers/contactDetail/contactDetail.reducer'
import { Ionicons } from '@expo/vector-icons'
import { ListItem, Avatar, SearchBar } from 'react-native-elements'
import { StatusBar } from 'expo-status-bar'

export const ContactListScreen = props => {

  const {
    navigation,
    setContactId,
    getContacts,
    deleteContact,
    setSearch,
    contactList,
    searchString,
    error,
  } = props;

  useEffect(() => {
    async function fetchData() {
      await getContacts();
    }
    fetchData();
  },[])

  const onDelete = async (contactId) => {
    await deleteContact(contactId)
    await getContacts();
  }

  const seeContactDetail = async (contactId) => {
    await setContactId(contactId)
    navigation.navigate('detail')
  }
    
  return (
    <View style={styles.container}>
      <SearchBar
        placeholder='search contact...'
        onChangeText={setSearch}
        onClear={()=> setSearch('')}
        value={searchString}
        searchIcon={{name: 'search', color:'#E88C30', type: 'font-awesome'}}
        clearIcon={{name: 'times', type: 'font-awesome'}}
        cancelIcon = {false}
        containerStyle = {styles.searchCont}
        inputContainerStyle = {styles.searchInput}
      />
        <FlatList
          data={ contactList }
          renderItem={({ item }) => 
            <ListItem
              bottomDivider
              onPress = {()=>seeContactDetail(item.id)}
            >
              <Avatar
                title={item.firstName.charAt(0)+item.lastName.charAt(0)}
                rounded
                size='medium'
                source={{uri: item.photo}}
              />
              <ListItem.Content
                style={{flex: 3}}
              >
                <ListItem.Title>{item.firstName+" "+item.lastName}</ListItem.Title>
              </ListItem.Content>
              <ListItem.Chevron
                name={'trash'}
                color={'red'}
                type={'font-awesome'}
                size={20}
                containerStyle={{flex:1}}
                onPress={()=> {
                  Alert.alert(
                    "Delete Contact",
                    "are you sure tou want to delete this contact?",
                    [
                      {
                        text: "cancel",
                        style: "cancel"
                      },
                      {
                        text: "OK",
                        onPress: async () => {
                          deleteContact(item.id)
                          await getContacts()
                        }
                      }
                    ]
                  )
                }}
              />
            </ListItem>
          }
          keyExtractor={item => item.id}
          
        />
        <View style={ styles.fabContainer }>
            <TouchableOpacity
            onPress = {()=>seeContactDetail('')}
            style={styles.fabButton}>
                <Ionicons name='ios-add' color='#fff' size={50} />
            </TouchableOpacity>
        </View>
    </View>
  )
}

const mapStateToProps = state => ({
  contactList: state.contactList.displayed,
  searchString: state.contactList.searchString,
  error: state.contactList.error,
});

const mapDispatchToProps = {
  getContacts: actions.fetchContactList,
  deleteContact: actions.deleteSelectedContact,
  setSearch: actions.setSearch,
  setContactId: detailActions.setContactId,
};

export default connect( mapStateToProps, mapDispatchToProps )(ContactListScreen);