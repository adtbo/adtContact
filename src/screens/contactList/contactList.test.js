import * as React from 'react';
import renderer from 'react-test-renderer';
import { ContactListScreen } from './contactList.screen';
import { ListItem, Avatar, SearchBar } from 'react-native-elements'
import { TouchableOpacity, FlatList } from 'react-native'

let props = {
    error: {},
    navigation: {
        navigate: jest.fn()
    },
    getContacts: jest.fn(() => Promise.resolve({})),
    setSearch: jest.fn(() => Promise.resolve({})),
    setContactId: jest.fn(() => Promise.resolve({})),
    deleteContac: jest.fn(),
    contactList: [],
    searchString: '',
};

let newProps = JSON.parse(JSON.stringify(props));
    newProps.contactList= [
    {firstName : "John",
    lastName: "Doe",
    photo: "some photo",
    age: 100},
    {firstName : "Jane",
    lastName: "Doe",
    photo: "some photo",
    age: 88},
    ]

describe('ContactList form Screen Unit test', () => {
  let tree;

  jest.useFakeTimers();

  it(`renders blank correctly`, async () => {
    await renderer.act(() => {
      tree = renderer.create(<ContactListScreen  { ...props} />).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it(`renders blank correctly`, async () => {
    await renderer.act(() => {
      tree = renderer.create(<ContactListScreen  { ...newProps} />).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  describe('check each component',() => {

    it('SearchBar', async () =>{
  
      await renderer.act(() => {
        component = renderer.create(<ContactListScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const srcBar = root.findAllByType(SearchBar)[0].props;
        await srcBar.onChangeText();
        expect(props.setSearch).toHaveBeenCalled();
      })
    })

    it('FlatList', async () =>{
  
      await renderer.act(() => {
        component = renderer.create(<ContactListScreen { ...newProps } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const list = root.findAllByType(FlatList)[0];
        //expect(list[0].props.length).toBe(1)
      })
    })

    it('TouchableOppacity', async () =>{
  
      await renderer.act(() => {
        component = renderer.create(<ContactListScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const button = root.findAllByType(TouchableOpacity)[0].props;
        await button.onPress();
        expect(props.navigation.navigate).toHaveBeenCalled();
      })
    })
  });
});