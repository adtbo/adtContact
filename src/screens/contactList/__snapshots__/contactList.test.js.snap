// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList form Screen Unit test renders blank correctly 1`] = `null`;

exports[`ContactList form Screen Unit test renders blank correctly 2`] = `null`;
